stake-pool/birdbath/src/processor.rs

```diff
@@ -0,0 +1,1100 @@
+        // the merge succeeding implicitly validates all properties of the user stake account
+
+        // merge the user stake account, which is preauthed to us, into the pool stake account
+        Self::stake_merge(
+            vote_account_address,
+            user_stake_info.clone(),
+            pool_authority_info.clone(),
+            bump_seed,
+            pool_stake_info.clone(),
+            clock_info.clone(),
+            stake_history_info.clone(),
+            stake_program_info.clone(),
+        )?;
+
+        let (_, post_validator_stake) = get_stake_state(pool_stake_info)?;
+        let post_validator_lamports = pool_stake_info.lamports();
+        msg!("Stake post merge {}", post_validator_stake.delegation.stake);
+
+        let lamports_added = post_validator_lamports
+            .checked_sub(pre_validator_lamports)
+            .ok_or(StakePoolError::CalculationFailure)?;
+
+        let stake_added = post_validator_stake
+            .delegation
+            .stake
+            .checked_sub(pre_validator_stake.delegation.stake)
+            .ok_or(StakePoolError::CalculationFailure)?;
+
+        let leftover_rent = lamports_added
+            .checked_sub(stake_added)
+            .ok_or(StakePoolError::CalculationFailure)?;
+
+        if stake_added != pre_user_stake.delegation.stake {
+            panic!("sanity check failed");
+        }
+
+        if leftover_rent != user_unstaked_lamports {
+            panic!("sanity check failed");
+        }
+
+        if user_stake_info.lamports() != 0 {
+            panic!("sanity check failed");
+        }
+
+        let token_supply = {
+            let pool_mint_data = pool_mint_info.try_borrow_data()?;
+            let pool_mint = StateWithExtensions::<Mint>::unpack(&pool_mint_data)?;
+            pool_mint.base.supply
+        };
+
+        let new_pool_tokens =
+            calculate_deposit_amount(token_supply, pre_validator_lamports, lamports_added)
+                .ok_or(StakePoolError::CalculationFailure)?;
+
+        if new_pool_tokens == 0 {
+            return Err(StakePoolError::DepositTooSmall.into());
+        }
+
+        // mint tokens to the user corresponding to their deposit
+        Self::token_mint_to(
+            vote_account_address,
+            token_program_info.clone(),
+            pool_mint_info.clone(),
+            user_token_account_info.clone(),
+            pool_authority_info.clone(),
+            bump_seed,
+            new_pool_tokens,
+        )?;
+
+        // return the lamports their stake account used to contain for rent-exemption
+        Self::stake_withdraw(
+            vote_account_address,
+            pool_stake_info.clone(),
+            pool_authority_info.clone(),
+            bump_seed,
+            user_lamport_account_info.clone(),
+            clock_info.clone(),
+            stake_history_info.clone(),
+            stake_program_info.clone(),
+            leftover_rent,
+        )?;
```

Comment:
This is extremely touchy logic -- in this case, by just sending the lamports back, it's possible for someone to take their rent.

During `merge`, the absorbed account gets set to `Uninitialized`, essentially zero'ed out, so there's no more data about who's authorized on the account. When the lamports get sent back, it's just an uninitialized stake with lamports, so anyone can call `initialize` on it to take it back. For them to keep their stake, we need to initialize it again after sending the lamports.

Reply:
this sends it back to the user wallet, not the uninitialized stake account. just in case, i made sure this actually works:
```
HANA lamps before staking: 999997955720
     lamps after staking: 998495657840 (1502297880 less than before, 2297880 excluding stake)
     lamps after deposit: 998497935720 (2277880 more than before)
```

Reply:
Oops, I misread the variables, my bad!